tripal/tripal
# Custom tables as base tables

# New Feature

Issue #177 

## Description


Users who create custom tables cannot use those tables as base tables.  This is not good because custom tables may be needed for new data types not currently supported by Chado.  Someone may create a custom module that creates custom tables and they may want to publish content from those tables.  

## Testing?





Test with the following:
1) Pull the branch and clear the cache.
2) Run the `drush updatedb` to perform the `tripal_custom_tables` table update.
3) Navigate to Admin > Tripal > Data Storage > Chado > Custom Tables and either edit an existing table or add a new one.  Be sure to check the 'Is this a base table?.  For my testing I just used the `blast_organism` table that comes with the Tripal Analysis Blast module.
4) Go to the Admin > Tripal > Data Storage > Chado > Semantic Web page, find the table name and add vocabulary terms to the fields.
5) Create a new content type that uses the custom table.
6) Publish records for the new content type.  You should have new pages!

The only weirdness with this fix is that I had to force new fields added to custom tables to have a `c_` prefix before all field names. This is because we cannot guarantee that custom tables are using the fields in the Chado style and our built-in Tripal fields may not work. For example.  If someone does use the `blast_organisms` table as a base table, and then use the `OBI:organism` term as the term for the `organism_name` field of that table, it will create a field named `c_obi__organism`.  This is because our `obi__organism` field expects an `organism_id` field and the custom table does not.  Therefore, any custom module that wants to use a custom table, will have to specifically override this.
